src/altimeter.rs

```diff
@@ -0,0 +1,273 @@
+        self.com.handle_dma_interrupt();
+    }
+
+    fn process_fifo_buffer(
+        data: [u8; BMP581_BUF_SIZE]
+    ) -> [PressureTemp; BMP581_FRAME_COUNT] {
+        data
+        .chunks(6)
+        .map(|x| x.split_at(3))
+        .map(|(pres, temp)| frame_to_reading(pres, temp))
+        .collect::<Vec<_, 16>>()
+        .into_array()
+        .unwrap()
+    }
+}
+
+impl I2CMasterWriteReadDMA for BMP581 {
+    unsafe fn write_read_dma(
+        &mut self,
+        addr: u8,
+        bytes: &[u8],
+        buf: &mut [u8],
+        callback: Option<I2cCompleteCallback>
+    ) -> Result<(), nb::Error<i2c::Error>> {
+        self.com.write_read_dma(addr, bytes, buf, callback)
+    }
+}
+
+impl AltimeterFifoDMA<BMP388_FRAME_COUNT, BMP388_BUF_SIZE> for BMP388Wrapper {
+    const FIFO_READ_REG: u8 = 0x14;
+    const ADDRESS: u8 = 0x77;
+
+    fn dma_interrupt(&mut self) {
+        self.bmp.com.handle_dma_interrupt();
+    }
+
+    fn process_fifo_buffer(
+        data: [u8; BMP388_BUF_SIZE]
+    ) -> [PressureTemp; BMP388_FRAME_COUNT] {
+        let mut i: usize = 0;
+        let mut output = Vec::<PressureTemp, BMP388_FRAME_COUNT>::new();
+
+        while i < BMP388_BUF_SIZE - 7 {
+            let header = data[i];
+            let fh_mode = (header >> 6) & 0b11;
+            let fh_param = (header >> 2) & 0b1111;
+            let p = (fh_param & 1) != 0;
+            let t = ((fh_param >> 2) & 1) != 0;
+            let s = ((fh_param >> 3) & 1) != 0;
+
+            // TODO: log invalid frames
+            match fh_mode {
+                0b01 => {
+                    // Config change or error
+                    i += 2;
+                },
+                0b10 => match (s, t, p) { // Data frame
+                    (true, false, false) => {
+                        // sensortime 
+                        i += 4;
+                    },
+                    (false, true, false) => {
+                        // temp only
+                        i += 4;
+                    },
+                    (false, false, true) => {
+                        // pressure only
+                        i += 4;
+                    },
+                    (false, true, true) => {
+                        // pressure & temp
+                        let t = &data[i+1 .. i+4];
+                        let p = &data[i+4 .. i+7];
+                        output.push(frame_to_reading(p, t)).unwrap();
+                        i += 7;
+                    },
+                    (false, false, false) => {
+                        // Empty
+                        i += 2;
+                    },
+                    _ => {
+                        // Invalid
+                        break;
+                    }
+                },
+                _ => {
+                    // Invalid
+                    break;
+                }
+            }
+        }
+
+
+        output.resize(BMP388_FRAME_COUNT, PressureTemp { pressure: 0, temperature: 0 }).unwrap();
```

Comment:
this is definitely not a good idea, we should just return a Vec if there's a chance we won't fully fill the buffer otherwise the stage detect code is going to be reading invalid data which would be bad :tm:

Reply:
done